File: docker/local/python/time-in-proc.py

import datetime, json, sys, traceback, csv, resource

from utils import *
from tools import *

# imported function
from function import function


cfg = json.load(open(sys.argv[1], 'r'))
repetitions = cfg['benchmark']['repetitions']
disable_gc = cfg['benchmark']['disable_gc']
input_data = cfg['input']

timedata = [0] * repetitions
os_times = [0] * repetitions
try:
    start = start_benchmarking(disable_gc)
    for i in range(0, repetitions):
        begin = datetime.datetime.now()
        begin_times = resource.getrusage(resource.RUSAGE_SELF)
        res = function.handler(input_data)
        end_times = resource.getrusage(resource.RUSAGE_SELF)
        stop = datetime.datetime.now()
        print(res, file = open(
                get_result_prefix(LOGS_DIR, 'output', 'txt'),
                'w'
            ))
        timedata[i] = [begin, stop]
        os_times[i] = [begin_times, end_times]
    end = stop_benchmarking()
except Exception as e:
    print('Exception caught!')
    print(e)
    traceback.print_exc()


result = get_result_prefix(RESULTS_DIR, cfg['benchmark']['name'], 'csv')
with open(result, 'w') as f:
    csv_writer = csv.writer(f)
    csv_writer.writerow(['#Seconds from epoch.microseconds; CPU times are in microseconds'])
    csv_writer.writerow(['Begin','End','Duration','User','Sys'])
    for i in range(0, len(timedata)):
        csv_writer.writerow([
                timedata[i][0].strftime('%s.%f'),
                timedata[i][1].strftime('%s.%f'),
                (timedata[i][1] - timedata[i][0]) /
                    datetime.timedelta(microseconds=1),
                (os_times[i][1].ru_utime - os_times[i][0].ru_utime) * 1e6,
                (os_times[i][1].ru_stime - os_times[i][0].ru_stime) * 1e6
            ])

experiment_data = {}
experiment_data['repetitions'] = repetitions
experiment_data['timestamps'] = process_timestamps(timedata)
experiment_data['start'] = str(start)
experiment_data['end'] = str(end)
print(json.dumps({'experiment': experiment_data, 'runtime': get_config()}, indent=2))

File: docker/local/nodejs/time-in-proc.js

const tools = require('./tools'),
    fs = require('fs'),
    strftime = require('strftime'),
    f = require('./function/function'),
    util = require('util');
const createCsvWriter = require('csv-writer').createArrayCsvWriter;

let cfg = JSON.parse(fs.readFileSync(process.argv[2]));
let repetitions = cfg.benchmark.repetitions;
let disable_gc = cfg.benchmark.disable_gc;
let input_data = cfg.input;
let timedata = new Array(repetitions);
process.on('unhandledRejection', r => console.log(r));

// Due to the async nature of nodejs, we use 'then' functionality
// of promise to make sure that we start a new instance only after finishing
// the previous one. There's no other option to achieve true waiting and we don't
// want to start multiple instances and let them work concurrently.
let measurer = async function(repetition, finish) {
  if (repetition < repetitions) {
    let begin_timestamp = Date.now();
    let begin = process.hrtime();
    let cpuTimeBegin = process.cpuUsage();
    let ret = f.handler(input_data);
    ret.then((res) => {
      let cpuTimeEnd = process.cpuUsage();
      let stop_timestamp = Date.now();
      let stop = process.hrtime(begin);
      let output_file = tools.get_result_prefix(tools.LOGS_DIR, 'output', 'txt');
      fs.writeFileSync(output_file, JSON.stringify(res));
      let userTime = cpuTimeEnd.user - cpuTimeBegin.user;
      let sysTime = cpuTimeEnd.system - cpuTimeBegin.system;
      timedata[repetition] = [begin_timestamp, stop_timestamp, stop[0]*1e6 + stop[1]/1e3, userTime, sysTime];
      measurer(repetition + 1, finish);
    },
    (reason) => {
      console.log('Function invocation failed!');
      console.log(reason);
      process.exit(1);
    }
    );
  } else{
    finish();
  }
}
start = tools.start_benchmarking();
measurer(0,
  () => {
    end = tools.stop_benchmarking();
    let result = tools.get_result_prefix(tools.RESULTS_DIR, cfg.benchmark.name, 'csv')
    let csvWriter = createCsvWriter({
          path: result,
          header: ['Begin','End','Duration','User','Sys']
    });
    for(let i = 0; i < repetitions; ++i) {
      timedata[i][0] = strftime('%s.%L', new Date(timedata[i][0]));
      timedata[i][1] = strftime('%s.%L', new Date(timedata[i][1]));
    }
    let p = csvWriter.writeRecords(timedata);
    p.then( () => {
      let reduce_array = timedata.map( x => { x.pop(); return x} );
      experiment_data = {
        repetitions: repetitions,
        start: start,
        end: end,
        timestamps: reduce_array
      }
      console.log( JSON.stringify({experiment: experiment_data, runtime: tools.get_config()}, null, 2) )
    });
  }
);
